realm/realm-library/src/objectServer/java/io/realm/internal/network/OkHttpAuthenticationServer.java

```diff
@@ -54,7 +55,9 @@ public AuthenticateResponse loginUser(SyncCredentials credentials, URL authentic
             String requestBody = AuthenticateRequest.userLogin(credentials).toJson();
             return authenticate(authenticationUrl, requestBody);
         } catch (Exception e) {
-            return AuthenticateResponse.from(new ObjectServerError(ErrorCode.UNKNOWN, e));
+            // ConnectException is recoverable (with exponential backoff)
+            ErrorCode errorCode = (e instanceof ConnectException) ? ErrorCode.IO_EXCEPTION : ErrorCode.UNKNOWN;
+            return AuthenticateResponse.from(new ObjectServerError(errorCode, e));
```

Comment:
Shouldn't we create a utility method which creates `AuthenticateResponse` from `Exception`?